ChronoVisor/src/KeeperRegistry.cpp

```diff
@@ -505,7 +823,237 @@ int KeeperRegistry::notifyKeepersOfStoryRecordingStop(std::vector<KeeperIdCard>
+    RecordingGroupId group_id = grapher_id_card.getGroupId();
+    ServiceId admin_service_id = reg_msg.getAdminServiceId();
+
+    std::lock_guard<std::mutex> lock(registryLock);
+    //re-check state after ther lock is aquired
+    if(is_shutting_down()) { return chronolog::CL_ERR_UNKNOWN; }
+
+    //find the group that keeper belongs to in the registry
+    auto group_iter = recordingGroups.find(group_id);
+    if(group_iter == recordingGroups.end())
+    {
+        auto insert_return =
+                recordingGroups.insert(std::pair<RecordingGroupId, RecordingGroup>(group_id, RecordingGroup(group_id)));
+        if(false == insert_return.second)
+        {
+            LOG_ERROR("[KeeperRegistry] keeper registration failed to find RecordingGroup {}", group_id);
+            return chronolog::CL_ERR_UNKNOWN;
+        }
+        else { group_iter = insert_return.first; }
+    }
+
+    RecordingGroup& recording_group = ((*group_iter).second);
+
+    // it is possible that the Registry still retains the record of the previous re-incarnation of the grapher process
+    // check for this case and clean up the leftover record...
+
+    std::stringstream id_string;
+    id_string << grapher_id_card;
+    if(recording_group.grapherProcess != nullptr)
+    {
+        // start delayed destruction for the lingering Adminclient to be safe...
+
+        chl::GrapherProcessEntry* grapher_process = recording_group.grapherProcess;
+        if(grapher_process->active)
+        {
+            // start delayed destruction for the lingering Adminclient to be safe...
+
+            std::time_t delayedExitTime = std::chrono::high_resolution_clock::to_time_t(
+                    std::chrono::high_resolution_clock::now() + std::chrono::seconds(delayedDataAdminExitSeconds));
+
+            recording_group.startDelayedGrapherExit(*grapher_process, delayedExitTime);
+        }
+        else
+        {
+            //check if any existing delayed exit grapher processes can be cleared...
+            std::time_t current_time =
+                    std::chrono::high_resolution_clock::to_time_t(std::chrono::high_resolution_clock::now());
+            recording_group.clearDelayedExitGrapher(*grapher_process, current_time);
+        }
+
+        if(grapher_process->delayedExitGrapherClients.empty())
+        {
+            LOG_INFO("[KeeperRegistry] registerGrapherProcess has destroyed old entry for grapher {}", id_string.str());
+            delete grapher_process;
+            recording_group.grapherProcess = nullptr;
+        }
+        else
+        {
+            LOG_INFO("[KeeperRegistry] registration for Grapher{} cant's proceed as previous grapherClient isn't yet "
+                     "dismantled",
+                     id_string.str());
+            return CL_ERR_UNKNOWN;
+        }
+    }
+
+    //create a client of the new grapher's DataStoreAdminService listenning at adminServiceId
+    std::string service_na_string("ofi+sockets://");
```

Comment:
Can we use `confManager.GRAPHER_CONF.DATA_STORE_ADMIN_SERVICE_CONF.PROTO_CONF`? or do you think we need to add a section in `VISOR_CONF` section for this as the other end of the service?
I think we can use `confManager.GRAPHER_CONF.DATA_STORE_ADMIN_SERVICE_CONF.PROTO_CONF` for now but add a TODO. When we refactor `ConfigurationManager` to split server and client side sections, we can add in the corresponding `VISOR_CONF` section for this.

Reply:
The same problem exists for communication protocol for Keeper's data store admin service although it is not a part of this PR.

Reply:
TODO: we need to add fabric name to the ServiceId, KeeperIdCard, and GrapherIdCard